### PR TITLE
formulae: use `generate_completions`

### PR DESCRIPTION
**Pull request by max-ae**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

needs https://github.com/Homebrew/brew/pull/13536

cc @MikeMcQuaid 

As requested in https://github.com/Homebrew/brew/pull/13536, here is a sample PR to showcase the use of the new `generate_completions` DSL.

This PR does not contain all formulae that would benefit from this, but only those that could be autocorrected by the RuboCop from https://github.com/Homebrew/brew/pull/13553 (with manual modifications to merge autodetected shells, since the Cop is not yet able to do that).
This means that this PR only contains formulae previously using the following pattern
```
(bash_completion/"foo").write Utils.safe_popen_read(bin/"foo", "completions", "bash")
```

Once the RuboCop is finalized, I will update this PR to also include formulae using something along the lines of 
```
output = Utils.safe_popen_read(bin/"foo", "completion", "bash")
(bash_completion/"foo").write output
```

But in the meantime, I think these selected formulae showcase the new DSL quite well 😄

### CODE REVIEW

File: Formula/homeassistant-cli.rb
Comment (Rylan12) [inline Formula/homeassistant-cli.rb:178]: Does this work? I think since `binary` defaults to `bin/base_name` this should be fine. Also, see https://github.com/Homebrew/brew/pull/13536#pullrequestreview-1037685543

```suggestion
    generate_completions(base_name: "hass-cli", shells: [:bash, :zsh])
```

File: Formula/mongodb-atlas-cli.rb
Comment (Rylan12) [inline Formula/mongodb-atlas-cli.rb:35]: Same as above

```suggestion
    generate_completions(base_name: "atlas")
```

File: Formula/vespa-cli.rb
Comment (Rylan12) [inline Formula/vespa-cli.rb:32]: Same as above

```suggestion
      generate_completions(base_name: "vespa")
```

File: Formula/arkade.rb
Comment (MikeMcQuaid) [inline Formula/arkade.rb:35]: I feel like `bin/"arkade"`, maybe even including `completion` should be a mandatory argument because otherwise this DSL doesn't make it obvious that it's going to be shelling out to three commands. Alternatively, `generate_completions_from_executable` or something might make this more obvious.

**Review by MikeMcQuaid — COMMENTED** [review]
This is looking great so far, nice work! Makes it much easier to understand what's going on.

**Comment by max-ae** [comment]
I have updated the previously changed formulae with the updated DSL and RuboCop. 
As soon as both are finalized, I will work on also adding all non autocorrectable formulae to this PR too.

File: Formula/arkade.rb
Comment (MikeMcQuaid) [inline Formula/arkade.rb:37]: I feel like something like:

```suggestion
    generate_completions_from_executable("arkade")
```

or 

```suggestion
    generate_completions_from_executable("arkade", "completion")
```

would be more in keeping with our existing DSLs. It should be more obvious what command is getting run, I think.

**Comment by max-ae** [comment]
This branch now contains all formulae that can benefit from the new syntax 😄 
The only exceptions I encountered were `hatch`, `awscli` and `pipenv`, otherwise no formula needs to use the `(bash_completion/"foo).write` syntax anymore 🎉

**Comment by max-ae** [comment]
For this to pass CI, this will need either a `long-timeout` or `syntax-only` label 😄 (@Rylan12)

**Comment by max-ae** [comment]
Cool, looks like this is good to go once https://github.com/Homebrew/brew/pull/13536 is in a tagged release!

**Comment by max-ae** [comment]
This can be merged now that 3.5.10 is out, no? 😃 (@Rylan12)

**Comment by Rylan12** [comment]
> This can be merged now that 3.5.10 is out, no? 😃 (@Rylan12)

Yep! Other maintainers can feel free to review and merge. I probably won't be able to take a look for a bit but will get to it in the next few days if no one else does first.

File: Formula/cdktf.rb
Comment (Rylan12) [inline Formula/cdktf.rb:27]: Isn't the `--help` still needed here?

File: Formula/circleci.rb
Comment (Rylan12) [inline Formula/circleci.rb:36]: Currently, the `basename` isn't the same for both shells. Bash uses `circleck` and zsh uses `circleci`

File: Formula/doitlive.rb
Comment (Rylan12) [inline Formula/doitlive.rb:57]:
```suggestion
    generate_completions_from_executable(libexec/"bin/doitlive", "completion", shells: [:bash, :zsh],
                                         shell_parameter_format: :none)
```

File: Formula/dvc.rb
Comment (Rylan12) [inline Formula/dvc.rb:703]: What about just doing this for consistency? (this works, right?)

```suggestion
    generate_completions_from_executable(bin/"dvc", "completion", "-s", shells: [:bash, :zsh])
```

File: Formula/earthly.rb
Comment (Rylan12) [inline Formula/earthly.rb:33]: What about this instead?

```suggestion
    generate_completions_from_executable(bin/"earthly", "bootstrap", "--source", shells: [:bash, :zsh])
```

File: Formula/faas-cli.rb
Comment (Rylan12) [inline Formula/faas-cli.rb:39]: Again, I think it's better to just pass `--shell` as an arg like this instead of relying on knowing the `shell_parameter_format` rules.

```suggestion
    generate_completions_from_executable(bin/"faas-cli", "completion", "--shell", shells: [:bash, :zsh])
```

File: Formula/gh.rb
Comment (Rylan12) [inline Formula/gh.rb:35]:
```suggestion
    generate_completions_from_executable(bin/"gh", "completion", "-s")
```

File: Formula/hatch.rb
Comment (Rylan12) [inline Formula/hatch.rb:181]: Let's not change this if we don't have to

**Review by Rylan12 — COMMENTED** [review]
I haven't made it through a ton of these but here are some comments

File: Formula/cdktf.rb
Comment (max-ae) [reply Formula/cdktf.rb:27]: No, I checked this locally, the issue is resolved.

File: Formula/circleci.rb
Comment (max-ae) [reply Formula/circleci.rb:36]: It works, nevertheless. Also, I suppose the `circleck` in the bash base name has been a typo when adding the completions previously.

File: Formula/doitlive.rb
Comment (max-ae) [reply Formula/doitlive.rb:57]: Of course, I missed that!

File: Formula/dvc.rb
Comment (max-ae) [reply Formula/dvc.rb:703]: Yes, this works. Looks better, my initial commit must have been before we made commands into a vararg.

File: Formula/hatch.rb
Comment (max-ae) [reply Formula/hatch.rb:181]: I changed it to satisfy the to-be-introduced cop:
As we don't support passing custom envs to the new DSL, the cop currently skips `#safe_popen_read` calls with custom envs. However, this only works if `(zsh_completion/"_foo").write` and the `#safe_popen_read` call are on the same line, as its difficult tracking and subsequently skipping something like `(zsh_completion/"_foo").write output`

**Comment by max-ae** [comment]
I went over all changed formulae again and, as suggested, changed custom `shell_parameter_format` args to extended `commands`, where applicable.

File: Formula/ipfs.rb
Comment (Rylan12) [inline Formula/ipfs.rb:35]: Do we need this?

```suggestion
    generate_completions_from_executable(bin/"ipfs", "commands", "completion", base_name: "ipfs-completion", shells: [:bash])
```

File: Formula/kubekey.rb
Comment (Rylan12) [inline Formula/kubekey.rb:29]:
```suggestion
    generate_completions_from_executable(bin/"kk", "completion", "--type", base_name: "kk", shells: [:bash, :zsh])
```

File: Formula/pandoc.rb
Comment (Rylan12) [inline Formula/pandoc.rb:28]:
```suggestion
                                         shells: [:bash], shell_parameter_format: :none)
```

**Review by Rylan12 — COMMENTED** [review]
Here's my final set of comments. Everything else looks good 👍

File: Formula/ipfs.rb
Comment (max-ae) [reply Formula/ipfs.rb:35]: No, the installed completion works without the change.

**Comment by max-ae** [comment]
> Here's my final set of comments. Everything else looks good 👍

Great! Thank you so much again for assisting in the whole process! 😄

File: Formula/ipfs.rb
Comment (Rylan12) [reply Formula/ipfs.rb:35]: How did you test this?

Comment (max-ae) [reply Formula/ipfs.rb:35]: I installed the formula from source from this PR and checked (in bash) if the completions where working.
(Note: if you want to reproduce this, the system bash on Monterey is too old for the completion script, you'll need brewed bash.)

**Review by Rylan12 — APPROVED** [review]
Thanks for all of the work you've put into this, @max-ae!

**Pull request closed without merging**
